operators/base_constraint.py

```diff
@@ -124,3 +130,33 @@ def draw(self, context: Context):
 
         for key in self.property_keys:
             layout.prop(self, key)
+
+    def exists(self, context, constraint_type=None) -> bool:
+        if hasattr(self, "entity2"):
+            new_dependencies = [i for i in [self.entity1, self.entity2, self.sketch] if i is not None]
+
+            if (isinstance(self.entity1, (SlvsPoint3D, SlvsLine3D)) or 
+                    isinstance(self.entity2, (SlvsPoint3D, SlvsLine3D))):
+                max_distance_dof = 3
+            elif ((isinstance(self.entity1, SlvsLine2D) and self.entity2 is None) or 
+                    isinstance(self.entity1, SlvsPoint2D) and isinstance(self.entity2, SlvsPoint2D)):
+                max_distance_dof = 2
+            else:
+                max_distance_dof = 1
+        else:
+            new_dependencies = [i for i in [self.entity1, self.sketch] if i is not None]
+            max_distance_dof = 1
+
+        distance_dof = 0
+        for c in context.scene.sketcher.constraints.all:
+            if isinstance(c, constraint_type):
+                if set(c.dependencies()) == set(new_dependencies):
+                    if constraint_type == SlvsDistance:
```

Comment:
It seems quite odd to have specific type checks in the base class, this should be done in the relevant sub-class (e.g. VIEW3D_OT_slvs_add_distance).

Reply:
We can add the whole part to VIEW3D_OT_slvs_add_distance the main would then look like this:
```python
 def main(self, context):
        if isinstance(self.entity1, SlvsLine2D) and self.entity2 is None:
            dependencies = self.entity1.dependencies()
            if (isinstance(dependencies[0], SlvsPoint2D) and 
                    isinstance(dependencies[1], SlvsPoint2D)):
                # for loop changes the values of self.entity1 and self.entity2
                # from a line entity to its endpoints
                for i in range(0, 2):
                    state_data = self.get_state_data(i)
                    state_data["hovered"] = -1
                    state_data["type"] = type(dependencies[i])
                    state_data["is_existing_entity"] = True
                    state_data["entity_index"] = dependencies[i].slvs_index
                self.next_state(context)

        if (isinstance(self.entity1, (SlvsPoint3D, SlvsLine3D)) or 
                isinstance(self.entity2, (SlvsPoint3D, SlvsLine3D))):
            max_constraints = 3
        elif ((isinstance(self.entity1, SlvsLine2D) and self.entity2 is None) or 
                isinstance(self.entity1, SlvsPoint2D) and isinstance(self.entity2, SlvsPoint2D)):
            max_constraints = 2
        else:
            max_constraints = 1

        if not self.exists(context, SlvsDistance, max_constraints):
            self.target = context.scene.sketcher.constraints.add_distance(
                self.entity1,
                self.entity2,
                sketch=self.sketch,
                init=not self.initialized,
                **self.get_settings(),
            )
        return super().main(context)
```

and the exists in the base class:
```python
    def exists(self, context, constraint_type=None, max_constraints=1) -> bool:
        if hasattr(self, "entity2"):
            new_dependencies = [i for i in [self.entity1, self.entity2, self.sketch] if i is not None]
        else:
            new_dependencies = [i for i in [self.entity1, self.sketch] if i is not None]

        constraint_counter = 0
        for c in context.scene.sketcher.constraints.all:
            if isinstance(c, constraint_type):
                if set(c.dependencies()) == set(new_dependencies):
                    constraint_counter += 1
                    if constraint_counter >= max_constraints:
                        return True

        return False
```

Does this all look better? It does to me...

Reply:
Yes that looks better